controller/contour/finalizer_test.go

```diff
@@ -0,0 +1,104 @@
+/*
+
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package contour
+
+import (
+	operatorv1alpha1 "github.com/projectcontour/contour-operator/api/v1alpha1"
+	"github.com/projectcontour/contour-operator/util/slice"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+const (
+	finalizerSuffix = "-finalizer"
+	testFinalizer   = contourFinalizer + finalizerSuffix
+)
+
+var _ = Describe("Run controller", func() {
+
+	Context("When creating a contour", func() {
+		It("Finalizer should be set", func() {
+			By("By creating a contour")
+
+			key := types.NamespacedName{
+				Name:      contourName + finalizerSuffix,
+				Namespace: operatorNamespace,
+			}
+
+			// Create a contour with ns delete set
+			created := &operatorv1alpha1.Contour{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: key.Namespace,
+					Name:      key.Name,
+				},
+				Spec: operatorv1alpha1.ContourSpec{
+					Namespace: operatorv1alpha1.NamespaceSpec{
+						RemoveOnDeletion: true,
+					},
+				},
+			}
+			Expect(k8sClient.Create(ctx, created)).Should(Succeed())
+
+			// Check contour finalizer
+			By("By finalizing the contour")
+			Eventually(func() bool {
+				f := &operatorv1alpha1.Contour{}
+				k8sClient.Get(ctx, key, f)
+				return slice.ContainsString(f.Finalizers, contourFinalizer)
+			}, timeout, interval).Should(Equal(true))
+
+			// Update finalizers
+			updated := &operatorv1alpha1.Contour{}
+			Expect(k8sClient.Get(ctx, key, updated)).Should(Succeed())
+			updated.Finalizers = append(updated.Finalizers, testFinalizer)
+			Expect(k8sClient.Update(ctx, updated)).Should(Succeed())
+
+			// Check contour finalizer still exists
+			By("Expecting the contour and test finalizers")
+			Eventually(func() bool {
+				f := &operatorv1alpha1.Contour{}
+				k8sClient.Get(ctx, key, f)
+				return slice.ContainsString(f.Finalizers, contourFinalizer) &&
+					slice.ContainsString(f.Finalizers, testFinalizer)
+			}, timeout, interval).Should(Equal(true))
```

Comment:
```suggestion
			Eventually(func() bool {
  				f := &operatorv1alpha1.Contour{}
  				k8sClient.Get(ctx, key, f)
  				return f.Finalizers
  			}, timeout, interval).Should(ContainElements([]string{contourFinalizer, testFinalizer}))
```

Reply:
I tried the following, but the test fails:
```
// Check contour finalizer still exists
By("Expecting the contour and test finalizers")
Eventually(func() []string {
	f := &operatorv1alpha1.Contour{}
	k8sClient.Get(ctx, key, f)
	return f.Finalizers
}, timeout, interval).Should(ContainElement([]string{testFinalizer, contourFinalizer}))
```

I did the same test against a running cluster and the finalizers are as expected:
```
kind: Contour
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"operator.projectcontour.io/v1alpha1","kind":"Contour","metadata":{"annotations":{},"name":"contour-sample","namespace":"default"},"spec":{}}
  creationTimestamp: "2020-10-02T01:13:54Z"
  finalizers:
  - contour.operator.projectcontour.io/finalizer
  - contour.operator.projectcontour.io/finalizer-test
<SNIP>
```

Reply:
I'm refactoring the test to remove the finalizer and ensure that it's added back by the operator.

Reply:
Did you try"ContainElement" or  "ContainElements"?